MiniDB/DataBase.cs

```diff
@@ -0,0 +1,1170 @@
+
+                    redoTransaction = new DBTransaction<T>()
+                    {
+                        TransactionType = TransactionType.Redo,
+                        Item_ID = last_transaction.Item_ID,
+                        Transacted_item = null,
+                        Changed_property = last_transaction.Changed_property,
+                        Property_new = last_transaction.Property_old,
+                        Property_old = last_transaction.Property_new
+                    };
+                }
+
+                // store Redo transaction            
+                this.Transactions_DB.Insert(0, redoTransaction);
+                last_transaction.Active = false;
+
+                // reregister changed handler
+                transactedItem.PropertyChangedExtended += this.DataBaseItem_PropertyChanged;
+                this.CollectionChanged += this.DataBase_CollectionChanged;
+
+                this._cacheDB();
+            }
+
+            this.OnItemChanged(transactedItem.ID);
+            this.PublicOnPropertyChanged(nameof(this.CanRedo));
+            this.PublicOnPropertyChanged(nameof(this.CanUndo));
+        }
+        #endregion
+
+        #region private methods
+        /// <summary>
+        /// Load in json from file
+        /// </summary>
+        /// <param name="filename">filename or path to read</param>
+        /// <returns>db as string to deserialize</returns>
+        protected virtual string _readFile(string filename)
+        {
+            if (System.IO.File.Exists(filename))
+            {
+                var json = System.IO.File.ReadAllText(filename);
+                return json;
+            }
+
+            return string.Empty;
+        }
+
+        /// <summary>
+        /// Get the transaction DB setup
+        /// </summary>
+        /// <param name="transactions_filename">file or path to where transactions DB should be stored</param>
+        /// <returns>new DB of DBTransaction of T</returns>
+        protected virtual DataBase<DBTransaction<T>> _getTransactionsDB(string transactions_filename)
+        {
+            return new DataBase<DBTransaction<T>>(transactions_filename, this.DBVersion, this.MinimumCompatibleVersion, true);
+        }
+
+        /// <summary>
+        /// Store the database to disk in <see cref="Filename"/>.
+        /// </summary>
+        protected virtual void _cacheDB()
+        {
+            lock (Locker)
+            {
+                var json = this.SerializeData;
+                System.IO.File.WriteAllText(this.Filename, json);
+            }
+        }
+
+        /// <summary>
+        /// Use reflection to set reverse the last transaction on the transacted item
+        /// </summary>
+        /// <param name="last_transaction">what happened that we're changing back</param>
+        /// <param name="transactedItem">the item to act on</param>
+        private static void SetProperty(DBTransaction<T> last_transaction, T transactedItem)
```

Comment:
Another very long method. A good rule of thumb is that a method is too long to fit on a single screen, then it likely should be refactored into smaller pieces.